### PR TITLE
Web Deployment

### PR DESCRIPTION
**Pull request by dmdashenkov**
## `local-web`

Introducing the `web` deployment configuration.

The deployment uses the servlet-based Spine `web` API and its Firebase implementation. The command and query endpoints are defined under the `/command` and `/query` paths respectively.

The `local-web` module uses the Gretty plugin to run locally. Run `./gradlew :local-web:appRun` to start the server.

### Prerequirements

The deployment uses the Firebase Realtime Database. It is required that the Firebase credentials are available under the `local-web/src/main/resources/spine-dev.json`.

To use a custom Firebase project (not `spine-dev`) put the credentials into the module classpath end edit the `io.spine.examples.todolist.server.FirebaseCredentials` class accordingly.

## `client-web`

Introducing the TodoList Web client.

Currently, the client is capable of creating and displaying simple tasks (i.e. tasks with a description only).

To run the client, open the `client-web/app/index.html` with the most convenient way (e.g. from IDEA).

### Prerequirements

In order to launch the Web application successfully
 - start the `local-web` server (see above);
 - if the server uses a custom Firebase project (not `spine-dev`), point the client to that project in `client-web/lib/firebase_client.js`;
 - build the `client-web` Gradle project with `./gradlew :client-web:build` (the general `./gradlew build` works as well).

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Comment by dmdashenkov** [comment]
@dpikhulya, @YegorUdovchenko, @PaulAgeyev, FYI. This is the examplar usage of the Spine `web` API (the JS part, in particular).

File: client-web/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Introducing the TodoList Web client.
+
+Currently, the client is capable of creating and displaying simple tasks (i.e. tasks with a description only).
+
+To run the client, open the `client-web/app/index.html` with the most convenient way (e.g. from IDEA).
+
+### Prerequirements
+
+In order to launch the Web application successfully
+ - start the `local-web` server (see above);
```
Comment (armiol) [inline client-web/README.md]: Where is that "above"? Please address.

File: client-web/build.gradle [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2018, TeamDev Ltd. All rights reserved.
```
Comment (armiol) [inline client-web/build.gradle]: Can you please strip all `Ltd`s from the codebase? See #63.

File: client/src/test/java/io/spine/examples/todolist/client/TodoClientTest.java [hunk 3]
```diff
@@ -70,6 +74,18 @@ void setUp() throws InterruptedException {
...
         getClient().shutdown();
+
+        // TODO:2018-05-10:dmytro.dashenkov: Remove when https://github.com/SpineEventEngine/core-java/issues/690 is resolved.
+
+        // TODO                              This code is deliberately ugly to make it obvious that
+        // TODO                              it should be removed ASAP.
+        try {
+            final Field field = ServerEnvironment.class.getDeclaredField("sharding");
+            field.setAccessible(true);
+            field.set(ServerEnvironment.getInstance(), new InProcessSharding(InMemoryTransportFactory.newInstance()));
```
Comment (armiol) [inline client/src/test/java/io/spine/examples/todolist/client/TodoClientTest.java]: Please see `ServerEnvironment.replaceSharding(Sharding)`.

File: deployment/local-web/src/main/java/io/spine/examples/todolist/server/CrossOriginResourceSharingFilter.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            header.appendTo(response);
+        }
+    }
+
+    /**
+     * The HTTP headers which configure the cross-origin request handling.
+     */
+    private enum ResponseHeader {
+
+        ACCESS_CONTROL_ALLOW_ORIGIN("access-control-allow-origin", "*"),
```
Comment (armiol) [inline deployment/local-web/src/main/java/io/spine/examples/todolist/server/CrossOriginResourceSharingFilter.java]: See https://google.github.io/guava/releases/19.0/api/docs/com/google/common/net/HttpHeaders.html.

File: deployment/local-web/src/main/java/io/spine/examples/todolist/server/FirebaseClient.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * A Firebase Realtime Database client.
+ *
+ * @author Dmytro Dashenkov
+ */
+final class FirebaseClient {
+
+    static {
+        try {
+            final InputStream in = FirebaseClient.class.getClassLoader()
```
Comment (armiol) [inline deployment/local-web/src/main/java/io/spine/examples/todolist/server/FirebaseClient.java]: Let's extract this portion to `FirebaseCredentials` and have it commented.

File: deployment/local-web/src/main/java/io/spine/examples/todolist/server/TodoQueryServlet.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                                                             .queryService())
+                                 .setDatabase(FirebaseClient.database())
+                                 .build());
+    }
+
+    @Override
+    protected void doOptions(HttpServletRequest req, HttpServletResponse resp) {
+    }
+}
+
```
Comment (armiol) [inline deployment/local-web/src/main/java/io/spine/examples/todolist/server/TodoQueryServlet.java]: There are some redundant empty lines here. Please remove.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: deployment/local-web/src/main/java/io/spine/examples/todolist/server/FirebaseClient.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * @author Dmytro Dashenkov
+ */
+final class FirebaseClient {
+
+    static {
+        final GoogleCredentials credentials = FirebaseCredentials.read();
+        final FirebaseOptions options = new FirebaseOptions.Builder()
+                .setCredentials(credentials)
+                .setDatabaseUrl("https://spine-dev.firebaseio.com/")
```
Comment (armiol) [inline deployment/local-web/src/main/java/io/spine/examples/todolist/server/FirebaseClient.java]: Let's have a constant with an explanation. I would even have a single `.gradle` file (such as `ext.gradle`) where all the stuff, that should be replaced, goes.

Right now it's virtually impossible for a third-party person to check out the project and run it — as she's not able to access any of our dev accounts.

Let's improve the configurability in the following PR. The aim is to allow people to run it in their cloud easily.

**Pull request merged**